lib/vmpooler/providers.rb

```diff
@@ -1,7 +1,116 @@
-%w[base dummy vsphere].each do |lib|
-  begin
-    require "vmpooler/providers/#{lib}"
-  rescue LoadError
-    require File.expand_path(File.join(File.dirname(__FILE__), 'providers', lib))
+require 'pathname'
+
+module Vmpooler
+  class Providers
+
+    # @param names [Array] - an array of names or string name of a provider
+    # @return [Array] - list of provider files loaded
+    # ie. ["lib/vmpooler/providers/base.rb", "lib/vmpooler/providers/dummy.rb", "lib/vmpooler/providers/vsphere.rb"]
+    def self.load_by_name(names)
+      names = Array(names)
+      instance = self.new
+      names.map {|name| instance.load_from_gems(name)}.flatten
+    end
+
+    # @return [Array] - array of provider files
+    # ie. ["lib/vmpooler/providers/base.rb", "lib/vmpooler/providers/dummy.rb", "lib/vmpooler/providers/vsphere.rb"]
+    # although these files can come from any gem
+    def self.load_all_providers
+      self.new.load_from_gems
+    end
+
+    # @return [Array] - returns an array of gem names that contain a provider
+    def self.installed_providers
+      self.new.vmpooler_provider_gem_list.map(&:name)
+    end
+
+    # @return [Array] returns a list of vmpooler providers gem plugin specs
+    def vmpooler_provider_gem_list
+      gemspecs.find_all { |spec| File.directory?(File.join(spec.full_gem_path, provider_path)) } + included_lib_dirs
+    end
+
+    # Internal: Find any gems containing vmpooler provider plugins and load the main file in them.
+    #
+    # @return [Array[String]] - a array of provider files
+    # @param name [String] - the name of the provider to load
+    def load_from_gems(name = nil)
```

Comment:
This already exists.  But the code is in the pool_manager.rb file instead because I need to separate the concerns.  The code below is what currently resides in this MR for pool_manager.

```
# load only providers used in config file
    def load_used_providers
      Vmpooler::Providers.load_by_name(used_providers)
    end

    # @return [Array] - a list of used providers from the config file, defaults to the default providers
    # ie. ["vsphere", "dummy"]
    def used_providers
      pools = config[:pools] || []
      @used_providers ||= (pools.map { |pool| pool[:provider] || pool['provider'] }.compact + default_providers ).uniq
    end

    # @param [Array] - returns a list of providers that should always be loaded
    # note: vsphere is the default if user does not specify although this should not be
    # if vsphere is to no longer be loaded by default please remove
    def default_providers
      @default_providers ||= %w( vsphere dummy )
    end
```

Reply:
Thanks @logicminds . I'm good with merging this, just have one last nitpick on the test name that is untitled and otherwise this looks ready to roll.

Reply:
Pretty sure I addressed that already.

```
it '#correct class' do
    expect(providers).to be_a Vmpooler::Providers
end
```

Reply:
In the providers spec I see your example, but the one I was referring to is the one with an in-line comment here https://github.com/puppetlabs/vmpooler/pull/263/files#diff-04094802b5bab73dcebfdd6e91f2b0a1R62